### PR TITLE
Feature/Generate Freeze- and Defrost Plans

### PR DESCRIPTION
**Pull request by st107755**
Automated resumption of scheduled TOSCA applications

Two new bpel plan builder added. 
They create new backup plans if the interface "http://opentosca.org/interfaces/backup" is added

- Start and end date:  25.04.18 - 25.10.18
- Supervisor: Lukas Harzenetter

### CODE REVIEW

File: org.opentosca.planbuilder.core.bpel/src/org/opentosca/planbuilder/core/bpel/BPELFreezeProcessBuilder.java [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+	 */
+	private List<BPELScopeActivity> runPlugins(final BPELPlan plan, final PropertyMap propMap) {
+
+		final List<BPELScopeActivity> changedActivities = new ArrayList<>();
+
+		for (final BPELScopeActivity templatePlan : plan.getTemplateBuildPlans()) {
+			if (templatePlan.getNodeTemplate() != null) {
+
+				// Only looks at Nodes that are no docker engine
+				if (!org.opentosca.container.core.tosca.convention.Utils
```
Comment (lharzenetter) [inline org.opentosca.planbuilder.core.bpel/src/org/opentosca/planbuilder/core/bpel/BPELFreezeProcessBuilder.java]: Why do you need this here?

File: org.opentosca.planbuilder.core.bpel/src/org/opentosca/planbuilder/core/bpel/BPELDefreezeProcessBuilder.java [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+	 */
+	private List<BPELScopeActivity> runPlugins(final BPELPlan plan, final PropertyMap propMap) {
+
+		final List<BPELScopeActivity> changedActivities = new ArrayList<>();
+
+		for (final BPELScopeActivity templatePlan : plan.getTemplateBuildPlans()) {
+			if (templatePlan.getNodeTemplate() != null) {
+
+				// Only looks at Nodes that are no docker engine
+				if (!org.opentosca.container.core.tosca.convention.Utils
```
Comment (lharzenetter) [inline org.opentosca.planbuilder.core.bpel/src/org/opentosca/planbuilder/core/bpel/BPELDefreezeProcessBuilder.java]: Why do you need this?

File: org.opentosca.planbuilder.core.bpel/src/org/opentosca/planbuilder/core/bpel/BPELFreezeProcessBuilder.java (hunk 1 above)
Comment (st107755) [reply org.opentosca.planbuilder.core.bpel/src/org/opentosca/planbuilder/core/bpel/BPELFreezeProcessBuilder.java]: The Docker Engine is the DinD Container. It's not good for the security if i allow to run scripts on this.
The script could mess with the other deployed Services

File: org.opentosca.planbuilder.core.bpel/src/org/opentosca/planbuilder/core/bpel/BPELDefreezeProcessBuilder.java [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+			return newTerminationPlan;
+		}
+
+		BPELDefreezeProcessBuilder.LOG.warn(
+				"Couldn't create DeFreezePlan for ServiceTemplate {} in Definitions {} of CSAR {}",
+				serviceTemplateId.toString(), definitions.getId(), csarName);
+		return null;
+	}
+
+	private boolean isDockerContainer(final AbstractNodeTemplate nodeTemplate) {
```
[the remark this answers is not in the document]
Comment (nyuuyn) [reply org.opentosca.planbuilder.core.bpel/src/org/opentosca/planbuilder/core/bpel/BPELDefreezeProcessBuilder.java]: removed

File: org.opentosca.planbuilder.core.bpel/src/org/opentosca/planbuilder/core/bpel/BPELFreezeProcessBuilder.java [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+				BPELFreezeProcessBuilder.LOG
+						.debug("Created BackupPlan " + newBuildPlan.getBpelProcessElement().getAttribute("name"));
+				plans.add(newBuildPlan);
+			}
+
+		}
+		return plans;
+	}
+
+	private boolean isDockerContainer(final AbstractNodeTemplate nodeTemplate) {
```
[the remark this answers is not in the document]
Comment (nyuuyn) [reply org.opentosca.planbuilder.core.bpel/src/org/opentosca/planbuilder/core/bpel/BPELFreezeProcessBuilder.java]: removed

File: org.opentosca.planbuilder.core.bpel/src/org/opentosca/planbuilder/core/bpel/BPELFreezeProcessBuilder.java [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+			 */
+
+			BPELFreezeProcessBuilder.LOG.debug("Created Plan:");
+			BPELFreezeProcessBuilder.LOG.debug(ModelUtils.getStringFromDoc(newTerminationPlan.getBpelDocument()));
+
+			return newTerminationPlan;
+		}
+
+		BPELFreezeProcessBuilder.LOG.warn(
+				"Couldn't create BackupPlan for ServiceTemplate {} in Definitions {} of CSAR {}",
```
[the remark this answers is not in the document]
Comment (nyuuyn) [reply org.opentosca.planbuilder.core.bpel/src/org/opentosca/planbuilder/core/bpel/BPELFreezeProcessBuilder.java]: fixed

**Comment by nyuuyn** [comment]
I worked on some of the change requests are there more I don't see them, i.e., do we need explicit "docker-support" from @st107755 ? Or is the abstract interface now okay? because we can basically just remove the old code

**Comment by lharzenetter** [comment]
works with `{http://opentosca.org/servicetemplates}MyTinyToDo-MySql_OpenStack-w2-wip1` at https://github.com/OpenTOSCA/tosca-definitions-internal/pull/50/commits/fbc90682381cbda14eb70ceee9109139a689b670

**Pull request merged**